pkg/controller/pinnedimageset/pinned_image_set.go

```diff
@@ -0,0 +1,399 @@
+	ctrl.enqueueMachineConfigPool(curPool)
+}
+
+func (ctrl *Controller) deleteMachineConfigPool(obj interface{}) {
+	pool, ok := obj.(*mcfgv1.MachineConfigPool)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("failed to get object from tombstone %#v", obj))
+			return
+		}
+		pool, ok = tombstone.Obj.(*mcfgv1.MachineConfigPool)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a MachineConfigPool %#v", obj))
+			return
+		}
+	}
+	klog.V(4).Infof("Deleting MachineConfigPool %s", pool.Name)
+}
+
+func (ctrl *Controller) addPinnedImageSet(obj interface{}) {
+	imageSet := obj.(*mcfgv1alpha1.PinnedImageSet)
+	if imageSet.DeletionTimestamp != nil {
+		ctrl.deletePinnedImageSet(imageSet)
+		return
+	}
+
+	pools, err := ctrl.getPoolsForPinnedImageSet(imageSet)
+	if err != nil {
+		klog.Errorf("error finding pools for pinned image set: %v", err)
+		return
+	}
+
+	klog.V(4).Infof("PinnedImageSet %s added", imageSet.Name)
+	for _, p := range pools {
+		ctrl.enqueueMachineConfigPool(p)
+	}
+}
+
+func (ctrl *Controller) updatePinnedImageSet(old, cur interface{}) {
+	oldImageSet := old.(*mcfgv1alpha1.PinnedImageSet)
+	newImageSet := cur.(*mcfgv1alpha1.PinnedImageSet)
+
+	pools, err := ctrl.getPoolsForPinnedImageSet(newImageSet)
+	if err != nil {
+		klog.Errorf("error finding pools for pinned image set: %v", err)
+		return
+	}
+
+	klog.V(4).Infof("PinnedImageSet %s updated", newImageSet.Name)
+	if triggerPinnedImageSetChange(oldImageSet, newImageSet) {
+		for _, p := range pools {
+			ctrl.enqueueMachineConfigPool(p)
+		}
+	}
+}
+
+func triggerPinnedImageSetChange(old, new *mcfgv1alpha1.PinnedImageSet) bool {
+	if old.DeletionTimestamp != new.DeletionTimestamp {
+		return true
+	}
+	if !reflect.DeepEqual(old.Spec, new.Spec) {
+		return true
+	}
+	return false
+}
+
+func (ctrl *Controller) deletePinnedImageSet(obj interface{}) {
+	imageSet, ok := obj.(*mcfgv1alpha1.PinnedImageSet)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("failed to get object from tombstone %#v", obj))
+			return
+		}
+		imageSet, ok = tombstone.Obj.(*mcfgv1alpha1.PinnedImageSet)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a PinnedImageSet %#v", obj))
+			return
+		}
+	}
+
+	pools, err := ctrl.getPoolsForPinnedImageSet(imageSet)
+	if err != nil {
+		klog.Errorf("error finding pools for pinned image set: %v", err)
+		return
+	}
+
+	klog.V(4).Infof("PinnedImageSet %s deleted", imageSet.Name)
+	for _, p := range pools {
+		ctrl.enqueueMachineConfigPool(p)
+	}
+}
+
+func (ctrl *Controller) getPoolsForPinnedImageSet(imageSet *mcfgv1alpha1.PinnedImageSet) ([]*mcfgv1.MachineConfigPool, error) {
+	if len(imageSet.Labels) == 0 {
+		return nil, fmt.Errorf("could not find any MachineConfigPool set for PinnedImageSet %s with labels: %v", imageSet.Name, imageSet.Labels)
```

Comment:
These errors seem to just get logged above. Would these populate to degrade something/

Reply:
good question should the controller go degraded if a CR has no labels.